crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is.md

```diff
@@ -0,0 +1,26 @@
+# Narrowing for `is` conditionals
+
+## `is None`
+
+```py
+x = None if flag else 1
+
+if x is None:
+    # TODO the following should be simplified to 'None'
+    reveal_type(x)  # revealed: None | Literal[1] & None
+
+reveal_type(x)  # revealed: None | Literal[1]
+```
+
+## `is` for other types
+
+```py
+x = 345
+y = x if flag else None
+
+if y is x:
+    # TODO the following should be simplified to 'Literal[345]'
+    reveal_type(y)  # revealed: Literal[345] | None & Literal[345]
```

Comment:
Hm, I understand why you would never want to do `if y is 345`, but doing `if y is x` (do these two variables refer to the exact same object) should be a valid — if obscure — use case?

I changed it to a custom class now, to make sure that this is not about something int-specific.

Reply:
_Is_ there a use case in Python for checking if two integer objects share the exact same address in memory (which is what the `is` operator does)? If you actually compare something to an integer _literal_, Python emits a `SyntaxWarning` to warn you that this is almost certainly not what you were trying to do, because you should nearly always treat one integer with value `42` the same as any other integer with value `42`:

```pycon
>>> x = 42
>>> x is 42
<python-input-1>:1: SyntaxWarning: "is" with 'int' literal. Did you mean "=="?
  x is 42
True
```

The `is` and `is not` operators are honestly sort of odd. Python attempts to abstract away the details of memory management in most cases, and for most objects it tries to discourage you from relying on details such as whether one object occupies the exact same memory address as another. And yet, the `id()` function is a builtin, and `is` and `is not` are crucial if you're using sentinel values such as `True`, `False`, `None` or `NotImplemented`.

Reply:
> _Is_ there a use case in Python for checking if two integer objects share the exact same address in memory (which is what the `is` operator does)?

I honestly don't know. But I could imagine some generic code, maybe in a serialization or caching library, that uses `x is y` to actually check whether two objects are the exact same thing. And that generic code might be used with simple integer objects as well. Anyway, I changed it to a custom class now.

Reply:
> But I could imagine some generic code, maybe in a serialization or caching library, that uses `x is y` to actually check whether two objects are the exact same thing.

Sounds like something I'd flag in a code review if I was helping maintain the library ;) anyway, using a custom class is definitely an improvement, thanks!

Reply:
Pytest doesn't have this but JS testing frameworks often expose a `expect(a).toBeSame(b)` and the function would do a `is` comparison internally.

Reply:
Yes, `unittest` has similar helper methods: `assertIs`, `assertIsNot`, `assertIsNone` and `assertIsNotNone`. (Pytest doesn't really need such helpers -- because of the AST rewriting it does before it runs the tests, just `assert x is y` will produce a beautiful traceback with lots of details if the test fails.) I'm not saying `is` comparisons are _useless_, just that `is` comparisons with immutable builtin types such as integers are almost always not really what you want to do, and that it's slightly strange that this functionality is exposed so prominently as builtin functions and syntax.